### PR TITLE
Add support for an explicit interface implementation methods

### PR DESCRIPTION
**Pull request by andriipatsula**
Issue: https://github.com/dotnet/arcade/issues/11895

* Add support to generate methods with `MethodKind.ExplicitInterfaceImplementation`
* Update common utility function `IsVisibleOutsideOfAssembly` to allow methods with visibility `Private` if they are of type ExplicitInterfaceImplementation

### CODE REVIEW

**Comment by andriipatsula** [comment]
@ViktorHofer , @smasher164 please review

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 1]
```diff
@@ -52,31 +52,63 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
...
+                            kinds: (tp.HasConstructorConstraint ? SpecialTypeConstraintKind.Constructor : SpecialTypeConstraintKind.None)
+                                   | (tp.HasReferenceTypeConstraint ? SpecialTypeConstraintKind.ReferenceType : SpecialTypeConstraintKind.None)
+                                   | (tp.HasValueTypeConstraint ? SpecialTypeConstraintKind.ValueType : SpecialTypeConstraintKind.None),
+                            types: tp.ConstraintTypes.Select(t => syntaxGenerator.TypeExpression(t)));
+                    }
                 }
             }
 
-            return syntaxGenerator.Declaration(symbol);
+            return declaration;
```
Comment (tkapin) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:111]: Why are we returning the SyntaxNode declaration when it's also an input parameter so it's changed anyway?

**Review by premun — APPROVED** [review]

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 2]
```diff
@@ -52,31 +52,63 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
...
-                                           | (tp.HasReferenceTypeConstraint ? SpecialTypeConstraintKind.ReferenceType : SpecialTypeConstraintKind.None)
-                                           | (tp.HasValueTypeConstraint ? SpecialTypeConstraintKind.ValueType : SpecialTypeConstraintKind.None),
-                                    types: tp.ConstraintTypes.Select(t => syntaxGenerator.TypeExpression(t)));
-                            }
-                        }
-                    }
+            return syntaxGenerator.Declaration(symbol);
+        }
+
+        // Temporary solution till corresponding Roslyn API is added: https://github.com/dotnet/arcade/issues/11895.
```
Comment (premun) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:72]: I'd add a "TODO"?

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 3]
```diff
@@ -52,31 +52,63 @@ public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, IS
...
+                decl = syntaxGenerator.WithTypeParametersAndConstraintsCopyExt(decl, method.TypeParameters);
+            }
+
+            return decl;
+        }
+
+        // this is copy/paste from private method `SyntaxGenerator.WithTypeParametersAndConstraints`
+        private static SyntaxNode WithTypeParametersAndConstraintsCopyExt(this SyntaxGenerator syntaxGenerator, SyntaxNode declaration, ImmutableArray<ITypeParameterSymbol> typeParameters)
+        {
+            if (typeParameters.Length > 0)
```
Comment (premun) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:94]: Can you invert the `if` and return early to reduce nesting?

**Review by premun — APPROVED** [review]

File: src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs [hunk 4]
```diff
@@ -56,6 +56,9 @@ public static IEnumerable<ITypeSymbol> GetAllBaseTypes(this ITypeSymbol type)
         public static bool IsEffectivelySealed(this ITypeSymbol type, bool includeInternals) =>
             type.IsSealed || !HasVisibleConstructor(type, includeInternals);
 
+        public static bool IsExplicitInterfaceImplementation(this ISymbol symbol) =>
+            symbol is IMethodSymbol method ? method.MethodKind == MethodKind.ExplicitInterfaceImplementation : false;
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs:60]:
```suggestion
        public static bool IsExplicitInterfaceImplementation(this ISymbol symbol) =>
            symbol is IMethodSymbol method && method.MethodKind == MethodKind.ExplicitInterfaceImplementation;
```

File: src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs [hunk 5]
```diff
@@ -91,6 +94,7 @@ symbol.DeclaredAccessibility switch
                 Accessibility.Protected => includeEffectivelyPrivateSymbols || symbol.ContainingType == null || !IsEffectivelySealed(symbol.ContainingType, includeInternals),
                 Accessibility.ProtectedOrInternal => includeEffectivelyPrivateSymbols || includeInternals || symbol.ContainingType == null || !IsEffectivelySealed(symbol.ContainingType, includeInternals),
                 Accessibility.ProtectedAndInternal => includeInternals && (includeEffectivelyPrivateSymbols || symbol.ContainingType == null || !IsEffectivelySealed(symbol.ContainingType, includeInternals)),
+                Accessibility.Private => includeEffectivelyPrivateSymbols && IsExplicitInterfaceImplementation(symbol),
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs:97]: I don't understand that change. Can you please elaborate? From what I see here, it looks wrong as private API isn't the same as an effectively private symbol (which is protected).

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 1 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:113]: It's like override of original `SyntaxGenerator.Declaration` methods for several use cases. If the use cases are not covered - call original `Declaration` method.

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 2 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: added

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 3 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: updated.

File: src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs (hunk 4 above)
Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs]: thanks

File: src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs [hunk 6]
```diff
@@ -56,6 +56,9 @@ public static IEnumerable<ITypeSymbol> GetAllBaseTypes(this ITypeSymbol type)
         public static bool IsEffectivelySealed(this ITypeSymbol type, bool includeInternals) =>
             type.IsSealed || !HasVisibleConstructor(type, includeInternals);
 
+        public static bool IsExplicitInterfaceImplementation(this ISymbol symbol) =>
```
[the remark this answers is not in the document]
Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs:64]: Updated

File: src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs (hunk 5 above)
Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs:102]:
```C#
using System;

public class MemoryManager : System.IDisposable
{
    /// you can call Dispose() only through pointer of `IDisposable` and not through pointer to `MemoryManager`
    void System.IDisposable.Dispose() { Console.WriteLine("1"); }
}

public class Test
{
	static void CallThroughIDisposableDirect(System.IDisposable mm)
	{
		mm.Dispose();
	}

	public static void Main()
	{
		MemoryManager mm = new MemoryManager();
		CallThroughIDisposableDirect(mm);
        // mm.Dispose(); <-- failure
	}
}
```

ExplicitInterfaceImplementation methods have `Accessibility.Private` in concrete example above.
Before, we filter out all private methods in `_ => includeInternals && symbol.DeclaredAccessibility != Accessibility.Private,`. if `includeEffectivelyPrivateSymbols` is false - we must (as I see this) provide just public interface..

Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs:102]: I forgot to mention, the the SyntaxGenerator does not support generation of ExplicitInterfaceImplementation methods and that's why I added support in the GenAPI. in concreate example above, SyntaxGenerator is generating
```C#
public class MemoryManager : System.IDisposable
{
     void System.IDisposable.Dispose() { Console.WriteLine("1"); }
}
```
-->
```
public class MemoryManager : System.IDisposable
{
}
```
as a result, compilation error.

Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/SymbolExtensions.cs:102]: Ok. I'll add new parameter `bool includeExplicitInterfaceImplementationSymbols = false`

File: src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/SymbolAccessibilityBasedFilter.cs [hunk 7]
```diff
@@ -2,22 +2,25 @@
...
 {
     public class SymbolAccessibilityBasedFilter : ISymbolFilter
     {
         private readonly bool _includeInternalSymbols;
         private readonly bool _includeEffectivelyPrivateSymbols;
+        private readonly bool _includeExplicitInterfaceImplementationSymbols;
 
-        public SymbolAccessibilityBasedFilter(bool includeInternalSymbols, bool includeEffectivelyPrivateSymbols = false)
+        public SymbolAccessibilityBasedFilter(bool includeInternalSymbols,
+            bool includeEffectivelyPrivateSymbols = false, bool includeExplicitInterfaceImplementationSymbols = false)
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/SymbolAccessibilityBasedFilter.cs:15]: nit: we usually declare either all or just one parameter per line

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs [hunk 8]
```diff
@@ -589,5 +589,28 @@ public partial class Car : System.IEquatable<Car>
...
+                        void System.IDisposable.Dispose() { }
+                    }
+                }
+                """,
+                expected: """
+                namespace Foo
+                {
+                    public abstract partial class MemoryManager : System.IDisposable
+                    {
+                        private void System.IDisposable.Dispose() { }
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:610]: Agreed. Can you please double check if that's valid C#?

[the remark this answers is not in the document]
Comment (andriipatsula) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:610]: Yep. just checked. `private` is not valid keyword (accessibility) here. Generation of ExplicitInterfaceImplementation methods is not implemented in SyntaxGenerator API. I tried to implement it in the GenAPI, but looks like I need to spend more time on this PR.

File: src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/SymbolAccessibilityBasedFilter.cs (hunk 7 above)
Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/SymbolAccessibilityBasedFilter.cs]: Updated

**Review by ViktorHofer — APPROVED** [review]
LGTM. Thanks

**Pull request merged**